consumer.go

```diff
@@ -513,13 +511,33 @@ retry:
 	if discoveryFilter, ok := r.behaviorDelegate.(DiscoveryFilter); ok {
 		nsqdAddrs = discoveryFilter.Filter(nsqdAddrs)
 	}
+
+	var successfulNsqdAddrs []string
 	for _, addr := range nsqdAddrs {
 		err = r.ConnectToNSQD(addr)
 		if err != nil && err != ErrAlreadyConnected {
 			r.log(LogLevelError, "(%s) error connecting to nsqd - %s", addr, err)
 			continue
 		}
+		successfulNsqdAddrs = append(successfulNsqdAddrs, addr)
 	}
+
+	// in the event that there are new nsqd addresses, remove the old connections from the connections map
+	for addr := range r.connections {
+		if !inAddrs(successfulNsqdAddrs, addr) {
```

Comment:
These successfulNsqdAddrs from a single lookupd. It is expected that the consumer effectively connects to the union of the results from multiple lookupd. Often, all lookupd return the same results, but it is expected that while changes to the set of nsqlookupd propagate there will be inconsistencies, and that's OK because of the "union" behavior. The behavior is also supposed to be resilient to some temporary network disruptions. So we don't really want to remove/disconnect nsqds which are missing from a single lookupd response.

Reply:
But the polling lookup loop only happens on the first lookupd address. https://github.com/nsqio/go-nsq/blob/b8bb2c58cd0829d58d66718d8d4dd6c925ca68b6/consumer.go#L383. We are having issues when the upstream IPs change and simply looking at the code it seemed troublesome that if the IPs change, old connections are not closed.

Reply:
When the first lookupd address is added, this `queryLookupd()` is called directly, then the `lookupdLoop()` is spawned to the background. That loop calls this `queryLookupd()` again periodically, while there usually are multiple lookupd addresses, and one of them is queried at a time by this function in a round-robin fashion.